pkg/util/permits/permits.go

```diff
@@ -0,0 +1,175 @@
+/*
+Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// package permits is used to provide permitGiver which maintains a sync map whose values can be
+// deleted if not accessed for a configured time
+package permits
+
+import (
+	"sync"
+	"time"
+
+	"k8s.io/klog/v2"
+)
+
+type PermitGiver interface {
+	RegisterPermits(key string, numPermits int)
+	TryPermit(key string, timeout time.Duration) bool
+	ReleasePermit(key string)
+	DeletePermits(key string)
+	Close()
+}
+
+type permit struct {
+	c                      chan struct{}
+	lastAcquiredPermitTime time.Time
+}
+
+func NewPermitGiver(stalePermitKeyTimeout time.Duration, janitorFreqency time.Duration) PermitGiver {
```

Comment:
Nit - Typo in the variable name `janitorFrequency`

Reply:
addressed in [59d31c0](https://github.com/gardener/machine-controller-manager/pull/683/commits/59d31c00569227a7e0466c1b4da4fe981f412562)